python/orca/tutorial/pytorch/NCF/train_data_loader.py

```diff
@@ -0,0 +1,237 @@
+    def __init__(self, features, 
+                num_item, train_mat=None, num_ng=0, is_training=None):
+        super(NCFData, self).__init__()
+        """ Note that the labels are only useful when training, we thus 
+            add them in the ng_sample() function.
+        """
+        self.features_ps = features
+        self.num_item = num_item
+        self.train_mat = train_mat
+        self.num_ng = num_ng
+        self.is_training = is_training
+        self.labels = [0 for _ in range(len(features))]
+
+    def ng_sample(self):
+        assert self.is_training, 'no need to sampling when testing'
+
+        self.features_ng = []
+        for x in self.features_ps:
+            u = x[0]
+            for t in range(self.num_ng):
+                j = np.random.randint(self.num_item)
+                while (u, j) in self.train_mat:
+                    j = np.random.randint(self.num_item)
+                self.features_ng.append([u, j])
+
+        labels_ps = [1 for _ in range(len(self.features_ps))]
+        labels_ng = [0 for _ in range(len(self.features_ng))]
+
+        self.features_fill = self.features_ps + self.features_ng
+        self.labels_fill = labels_ps + labels_ng
+
+    def __len__(self):
+        return (self.num_ng + 1) * len(self.labels)
+    
+    def __getitem__(self, idx):
+        features =  self.features_fill if self.is_training else self.features_ps
+        labels =  self.labels_fill if self.is_training else self.labels
+
+        user = features[idx][0]
+        item = features[idx][1]
+        label = float(labels[idx])
+        return [user, item] ,label
+
+# prepare the train and test datasets
+train_data, test_data, args.user_num, args.item_num, train_mat = load_all()
+
+# construct the train and test dataloader
+train_dataset = NCFData(
+        train_data, args.item_num, train_mat,args.num_ng, True)
+test_dataset = NCFData(
+        test_data, args.item_num, train_mat, 0, False)
+train_loader = data.DataLoader(train_dataset,
+        batch_size=args.batch_size, shuffle=True, num_workers=0)
+test_loader = data.DataLoader(test_dataset,
+        batch_size=args.test_num_ng+1, shuffle=False, num_workers=0)
+
+#Step 3: Define the Model
+
+# create the model
+def model_creator(config):
+    model = NCF(args.user_num, args.item_num,args.factor_num, args.num_layers, args.dropout, args.model) # a torch.nn.Module
+    model.train()
+    return model
+
+#create the optimizer
+def optimizer_creator(model, config):
+    return optim.Adam(model.parameters(), lr= args.lr)
+
+#define the loss function
+loss_function = nn.BCEWithLogitsLoss()
+
+#Step 4: Fit with Orca Estimator
+
+from bigdl.orca.learn.pytorch import Estimator 
+from bigdl.orca.learn.metrics import Accuracy
+
+# create the estimator
+est = Estimator.from_torch(model=model_creator, optimizer=optimizer_creator,loss=loss_function, metrics=[Accuracy()],backend=args.backend)# backend="ray" or "spark"
+
+def train_loader_func(config, batch_size):

```

Comment:
> An instance of SparkXShards, a Ray Dataset, a Spark DataFrame or a function
>                that takes config and batch_size as argument and returns a PyTorch DataLoader for
>                training.

I think here we do not have to use a function.

Reply:
??